ext/standard/random.c

```diff
@@ -0,0 +1,186 @@
+   | Authors: Sammy Kaye Powers <me@sammyk.me>                            |
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <math.h>
+
+#include "php.h"
+#include "php_random.h"
+
+#if PHP_WIN32
+# include "win32/winutil.h"
+#endif
+
+ZEND_DECLARE_MODULE_GLOBALS(random);
+
+/* {{{ */
+PHP_MINIT_FUNCTION(random)
+{
+	RANDOM_G(fd) = -1;
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ */
+static int php_random_bytes(void *bytes, size_t size)
+{
+#if PHP_WIN32
+	/* Defer to CryptGenRandom on Windows */
+	if (php_win32_get_random_bytes(bytes, size) == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "Could not gather sufficient random data");
+		return FAILURE;
+	}
+#else
+#if HAVE_DECL_ARC4RANDOM_BUF
+	arc4random_buf(bytes, size);
+#else
+	int    fd = RANDOM_G(fd);
+	size_t read_bytes = 0;
+
+	if (fd < 0) {
+#if HAVE_DEV_ARANDOM
+		fd = open("/dev/arandom", O_RDONLY);
+#else
+#if HAVE_DEV_URANDOM
+		fd = open("/dev/urandom", O_RDONLY);
+#endif // URANDOM
+#endif // ARANDOM
+		if (fd < 0) {
+			php_error_docref(NULL, E_WARNING, "Cannot open source device");
+			return FAILURE;
+		}
+
+		RANDOM_G(fd) = fd;
+	}
+
+	size_t n = 0;
+	while (read_bytes < size) {
+		n = read(fd, bytes + read_bytes, size - read_bytes);
+		if (n < 0) {
+			break;
+		}
+		read_bytes += n;
+	}
+
+	if (read_bytes < size) {
+		php_error_docref(NULL, E_WARNING, "Could not gather sufficient random data");
+		return FAILURE;
+	}
+#endif // !ARC4RANDOM_BUF
+#endif // !WIN32
+
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ proto string random_bytes(int length)
+Return an arbitrary length of pseudo-random bytes as binary string */
+PHP_FUNCTION(random_bytes)
+{
+	zend_long size;
+	zend_string *bytes;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &size) == FAILURE) {
+		return;
+	}
+
+	if (size < 1) {
+		php_error_docref(NULL, E_WARNING, "Length must be greater than 0");
+		RETURN_FALSE;
+	}
+
+	bytes = zend_string_alloc(size, 0);
+
+	if (php_random_bytes(bytes->val, size) == FAILURE) {
+		zend_string_release(bytes);
+		RETURN_FALSE;
+	}
+
+	bytes->val[size] = '\0';
+
+	RETURN_STR(bytes);
+}
+/* }}} */
+
+/* {{{ proto int random_int(int min, int max)
+Return an arbitrary pseudo-random integer */
+PHP_FUNCTION(random_int)
+{
+	zend_long min = ZEND_LONG_MIN;
+	zend_long max = ZEND_LONG_MAX;
+	zend_ulong limit;
+	zend_ulong umax;
+	zend_ulong result;
+
+	if (ZEND_NUM_ARGS() == 1) {
+		php_error_docref(NULL, E_WARNING, "A minimum and maximum value are expected, only minimum given");
+		RETURN_FALSE;
+	}
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|ll", &min, &max) == FAILURE) {
+		return;
+	}
+
+	if (min >= max) {
+		php_error_docref(NULL, E_WARNING, "Minimum value must be less than the maximum value");
+		RETURN_FALSE;
+	}
+
+	umax = max - min;
+
+	if (php_random_bytes(&result, sizeof(result)) == FAILURE) {
```

Comment:
@weltling There's no need for any extra work. If an integer occupies 64 bits of memory, and we use a random source to set every one of those bits, the result is a random integer with the same quality as the random source.

Endianness does not matter since every byte is independently random. If the bytes are ordered AB they are equally random to being ordered BA.

There are tools that can test the quality of the random output, but you'll literally be testing the underlying source. We're putting our faith in the Linux/Windows/BSD APIs here. If it turns out that these sources are in fact low quality, then civilisation itself will collapse :)